connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala

```diff
@@ -0,0 +1,198 @@
+      out => if (isDebug) Source.fromInputStream(out).getLines.foreach(debug),
+      err => if (isDebug) Source.fromInputStream(err).getLines.foreach(debug))
+    val process = builder.run(io)
+
+    // Adding JVM shutdown hook
+    sys.addShutdownHook(kill())
+    process
+  }
+
+  def start(): Unit = {
+    assert(!stopped)
+    sparkConnect
+  }
+
+  def kill(): Int = {
+    stopped = true
+    debug("Stopping the Spark Connect Server...")
+    sparkConnect.destroy()
+    val code = sparkConnect.exitValue()
+    debug(s"Spark Connect Server is stopped with exit code: $code")
+    code
+  }
+
+  private def fileSparkHome(): String = {
+    val path = new File("./").getCanonicalPath
+    if (path.endsWith("connector/connect/client/jvm")) {
+      // the current folder is the client project folder
+      new File("../../../../").getCanonicalPath
+    } else {
+      path
+    }
+  }
+
+  private def findSparkConnectJar: String = {
+    val target = "connector/connect/server/target"
+    val parentDir = new File(sparkHome, target)
+    assert(
+      parentDir.exists(),
+      s"Fail to locate the spark connect target folder: '${parentDir.getCanonicalPath}'. " +
+        s"SPARK_HOME='${new File(sparkHome).getCanonicalPath}'. " +
+        "Make sure system property `SPARK_HOME` is set correctly.")
+    val jars = recursiveListFiles(parentDir).filter { f =>
+      // SBT jar
+      (f.getParent.endsWith("scala-2.12") &&
+        f.getName.startsWith("spark-connect-assembly") && f.getName.endsWith("SNAPSHOT.jar")) ||
+      // Maven Jar
+      (f.getParent.endsWith("target") &&
+        f.getName.startsWith("spark-connect") && f.getName.endsWith("SNAPSHOT.jar"))
+    }
+    // It is possible we found more than one: one built by maven, and another by SBT
+    assert(
+      jars.nonEmpty,
+      s"Failed to find the `spark-connect` jar inside folder: ${parentDir.getCanonicalPath}")
+    debug("Using jar: " + jars(0).getCanonicalPath)
+    jars(0).getCanonicalPath // return the first one
+  }
+
+  def recursiveListFiles(f: File): Array[File] = {
+    val these = f.listFiles
+    these ++ these.filter(_.isDirectory).flatMap(recursiveListFiles)
+  }
+}
+
+trait RemoteSparkSession
+    extends org.scalatest.funsuite.AnyFunSuite // scalastyle:ignore funsuite
+    with BeforeAndAfterAll {
+  import SparkConnectServerUtils._
+  var spark: SparkSession = _
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    SparkConnectServerUtils.start()
+    spark = SparkSession.builder().client(SparkConnectClient.builder().port(port).build()).build()
+
+    // Retry and wait for the server to start
+    val stop = System.currentTimeMillis() + 60000 * 1 // ~1 min
+    var sleepInternal = 1000 // 1s with * 2 backoff
+    var success = false
+    val error = new RuntimeException(s"Failed to start the test server on port $port.")
+
+    while (!success && System.currentTimeMillis() < stop) {
```

Comment:
should actually avoid `currentTimeMillis` https://github.com/databricks/scala-style-guide#prefer-nanotime-over-currenttimemillis